pkg/changelog/changelog.go

```diff
@@ -6,219 +6,202 @@ import (
 	"github.com/aquasecurity/tracee/pkg/logger"
 )
 
-type comparable interface {
-	~int | ~float64 | ~string
-}
-
-type item[T comparable] struct {
-	timestamp time.Time // timestamp of the change
-	value     T         // value of the change
-}
+//
+// Entries
+//
 
-// The changelog package provides a changelog data structure. It is a list of changes, each with a
-// timestamp. The changelog can be queried for the value at a given time.
+// MemberKind represents the unique identifier for each kind of entry in the Entries.
+// It is used to categorize different kinds of changes tracked by the Entries.
+type MemberKind uint8
 
-// ATTENTION: You should use Changelog within a struct and provide methods to access it,
-// coordinating access through your struct mutexes. DO NOT EXPOSE the changelog object directly to
-// the outside world as it is not thread-safe.
+// MaxEntries represents the maximum number of entries that can be stored for a given kind of entry.
+type MaxEntries uint8
 
-type Changelog[T comparable] struct {
-	changes    []item[T]              // list of changes
-	timestamps map[time.Time]struct{} // set of timestamps (used to avoid duplicates)
-	maxSize    int                    // maximum amount of changes to keep track of
+// entry is an internal structure representing a single change in the Entries.
+// It includes the kind of the entry, the timestamp of the change, and the value of the change.
+type entry[T comparable] struct {
+	k     MemberKind // Kind of the member, used to categorize the entry.
+	t     time.Time  // Timestamp of when the change occurred.
+	value T          // Value of the change.
 }
 
-// NewChangelog creates a new changelog.
-func NewChangelog[T comparable](maxSize int) *Changelog[T] {
-	return &Changelog[T]{
-		changes:    []item[T]{},
-		timestamps: map[time.Time]struct{}{},
-		maxSize:    maxSize,
-	}
+// Entries is the main structure that manages a list of changes (entries).
+// It keeps track of specifically configured members indicated by MemberKind identifiers.
+// When instantiating an Entries struct, one must supply a relevant mapping between the desired
+// unique members and the maximum amount of changes that member can track.
+//
+// ATTENTION: You should use Entries within a struct and provide methods to access it,
+// coordinating access through your struct mutexes. DO NOT EXPOSE the Entries object directly to
+// the outside world as it is not thread-safe.
+type Entries[T comparable] struct {
+	entryFlags []MaxEntries // Configuration slice defining flags for each member kind.
+	entries    []entry[T]   // List of recorded entries.
```

Comment:
entries is tightly coupled to entryFlags. This can be simplified to:
```
type Entries[T comparable] struct {
    entries []struct {
        maxEntries MaxEntries
        entries    []entry[T]
    }
}
```

Also consider using slice of slices and remove `MemberKind` from `entry`

Reply:
Ok, I'm right now implementing the changes above in a linked list. Soon gonna get results after stress.

```go
type entry[T comparable] struct {
	time  time.Time // Timestamp of the change.
	value T         // Value of the change.
	prev  *entry[T] // Pointer to the previous entry in the linked list.
}

type entries[T comparable] struct {
	maxEntries MaxEntries // Maximum number of entries for this kind.
	size       uint8      // Current number of entries for this kind.
	tail       *entry[T]  // Tail pointer for the linked list of this kind.
}
```